### PR TITLE
[GraphBolt] Update `ItemSampler`

### PR DESCRIPTION
**Pull request by Skeleton003**
## Description

1. Update `ItemSampler` to support correct stochastic sharding across distributed groups.
2. Modify the logic of `ItemSet.__getitem__()` when index is an iterable of int.

Benchmark: https://docs.google.com/document/d/1Pzk2PJoFtTZSu17wTXVK4mqvfrMLAj2xK6fcGC1pwEg/edit?usp=sharing 

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Comment by mfbalin** [comment]
The fact that runtime performance is unchanged is good. However, to verify whether the old or the new implementation is more performant, we need to track the CPU utilization. Since ItemSampler and rest of the sampling pipeline runs concurrently, runtime is not enough information to determine that.

**Comment by Skeleton003** [comment]
> The fact that runtime performance is unchanged is good. However, to verify whether the old or the new implementation is more performant, we need to track the CPU utilization. Since ItemSampler and rest of the sampling pipeline runs concurrently, runtime is not enough information to determine that.

Sure, could u please give me some guide on how to do that?

**Comment by mfbalin** [comment]
> > The fact that runtime performance is unchanged is good. However, to verify whether the old or the new implementation is more performant, we need to track the CPU utilization. Since ItemSampler and rest of the sampling pipeline runs concurrently, runtime is not enough information to determine that.
> 
> Sure, could u please give me some guide on how to do that?

The simplest way to monitor `htop -d5` or `nvtop` to see what is the CPU utilization. If you wanted to be more precise, you could also insert timing code into ItemSampler to see how long each call takes. It may also make sense to write a regression benchmark so that we can monitor its runtime.

File: python/dgl/graphbolt/itemset.py [hunk 1]
```diff
@@ -362,33 +362,19 @@ def __getitem__(self, index: Union[int, slice, Iterable[int]]):
...
-                if len(indices) == 0:
-                    del data[key]
+            index = torch.tensor(index).squeeze()
+            assert torch.all((index >= 0) & (index < self._length))
+            key_indices = (
+                torch.searchsorted(self._offsets, index, right=True) - 1
+            )
+            data = {}
+            for key_id, key in enumerate(self._keys):
+                mask = (key_indices == key_id).nonzero().squeeze(1)
```
Comment (mfbalin) [inline python/dgl/graphbolt/itemset.py:372]: Is there a way to make this part more efficient?

Comment (mfbalin) [reply python/dgl/graphbolt/itemset.py:372]: If there are K etypes, then the runtime is `O(K * len(index))`. If there are a lot of etypes, it could easily become the bottleneck.

File: python/dgl/graphbolt/item_sampler.py [hunk 2]
```diff
@@ -510,68 +313,94 @@ def __init__(
...
         self._world_size = None
         self._rank = None
-        self._rng = np.random.default_rng()
-
-    def _organize_items(self, data_pipe) -> None:
-        # Shuffle before batch.
-        if self._shuffle:
-            data_pipe = data_pipe.shuffle(buffer_size=self._buffer_size)
+        self._seed = np.random.randint(0, np.iinfo(np.int32).max)
+        self._epoch = 0
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/item_sampler.py:317]: Add comment here to clarify why we need such an variable.

File: python/dgl/graphbolt/item_sampler.py [hunk 3]
```diff
@@ -756,6 +582,18 @@ def __init__(
             )
         self._world_size = dist.get_world_size()
         self._rank = dist.get_rank()
+        device = (
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/item_sampler.py:585]: how about wrapping this logic into a method as it's dedicated for seed initialization.

File: python/dgl/graphbolt/item_sampler.py [hunk 4]
```diff
@@ -510,68 +313,94 @@ def __init__(
...
+            yield self._minibatcher(
+                self._collate_batch(buffer, indices, offsets),
+                self._names,
+            )
 
-        return iter(data_pipe)
+        self._epoch += 1
 
 
 class DistributedItemSampler(ItemSampler):
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/item_sampler.py:406]: does it work well for `num_workers>0`?

File: python/dgl/graphbolt/item_sampler.py [hunk 5]
```diff
@@ -756,6 +582,18 @@ def __init__(
...
+            if torch.cuda.is_available() and dist.get_backend() == "nccl"
+            else "cpu"
+        )
+        if self._rank == 0:
+            seed = np.random.randint(0, np.iinfo(np.int32).max)
+            seed_tensor = torch.tensor(seed, dtype=torch.int32, device=device)
+        else:
+            seed_tensor = torch.empty([], dtype=torch.int32, device=device)
+        dist.broadcast(seed_tensor, src=0)
+        self._seed = seed_tensor.item()
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/item_sampler.py:596]: is it possible to add a dedicated testcase to make sure the collated minibatches are shuffled globally?

**Review by Rhett-Ying — COMMENTED** [review]
code looks clean and code to me.

One more thing is please benchmark on larger dataset with larger batch-size such as `ogbn-papers100M` and heterogenous dataset,  link prediction datasets(to measure the perf of indexing on tuple of tensors). Let's make sure it's performance efficient on most common datasets.

File: python/dgl/graphbolt/item_sampler.py (hunk 4 above)
Comment (Skeleton003) [reply python/dgl/graphbolt/item_sampler.py:406]: No, the issue #7381 still exist. I'll continue to investigate.

File: python/dgl/graphbolt/item_sampler.py (hunk 5 above)
Comment (Skeleton003) [reply python/dgl/graphbolt/item_sampler.py:596]: I'm concerned that since the shuffling are completely random, it is possible that the result will turn out to look like a local shuffling (like how we did before). It is possible to fail such a test, although the probability is very low.

File: python/dgl/graphbolt/itemset.py (hunk 1 above)
Comment (Skeleton003) [reply python/dgl/graphbolt/itemset.py:372]: This code is from https://github.com/dmlc/dgl/blob/5b3496892c16ee1bc80430baa50fc80bc1d425e7/python/dgl/graphbolt/item_sampler.py#L194-L207 
and is what we are currently using. Do you have any suggestions on how to improve it?

File: python/dgl/graphbolt/itemset.py [hunk 6]
```diff
@@ -362,33 +362,19 @@ def __getitem__(self, index: Union[int, slice, Iterable[int]]):
...
-                offset_idx = torch.searchsorted(self._offsets, idx, right=True)
-                offset_idx -= 1
-                idx -= self._offsets[offset_idx]
-                key = self._keys[offset_idx]
-                data[key].append(int(idx))
-            for key in self._keys:
-                indices = data[key]
-                if len(indices) == 0:
-                    del data[key]
+            index = torch.tensor(index).squeeze()
```
Comment (mfbalin) [inline python/dgl/graphbolt/itemset.py:365]:
```suggestion
            index = torch.tensor(index).squeeze().sort()[0]
```

File: python/dgl/graphbolt/itemset.py [hunk 7]
```diff
@@ -362,33 +362,19 @@ def __getitem__(self, index: Union[int, slice, Iterable[int]]):
...
+            assert torch.all((index >= 0) & (index < self._length))
+            key_indices = (
+                torch.searchsorted(self._offsets, index, right=True) - 1
+            )
```
Comment (mfbalin) [inline python/dgl/graphbolt/itemset.py:369]:
```suggestion
            index = torch.tensor(index).squeeze().sort()[0]
            assert index[0] >= 0 and index[-1] < self._length
            offsets = (
                torch.searchsorted(index, self._offsets, right=False)
            )
```

File: python/dgl/graphbolt/itemset.py (hunk 1 above)
Comment (mfbalin) [inline python/dgl/graphbolt/itemset.py:372]: here, we can use the offsets above to get which keys we need to process.

Comment (Skeleton003) [reply python/dgl/graphbolt/itemset.py:372]: That's brilliant!

File: python/dgl/graphbolt/itemset.py (hunk 7 above)
Comment (Skeleton003) [reply python/dgl/graphbolt/itemset.py:369]: Is it necessary to sort the index? What if we want to reserve the order?

Comment (mfbalin) [reply python/dgl/graphbolt/itemset.py:369]: We can then use the second output of sort to permute it back.

File: python/dgl/graphbolt/item_sampler.py (hunk 5 above)
Comment (Rhett-Ying) [reply python/dgl/graphbolt/item_sampler.py:603]: then please check if the model accuracy go up or down.

Comment (Skeleton003) [reply python/dgl/graphbolt/item_sampler.py:603]: Benchmark result updated in the Description. The performance of most examples remain the same.

**Comment by Skeleton003** [comment]
@mfbalin Thank you for your valuable suggestions, but considering the scope of this PR, I'd like to defer them to another PR. 

@Rhett-Ying If everything looks good to you, please approve so I can work on.

**Comment by Rhett-Ying** [comment]
@Skeleton003 According to the doc, `examples/sampling/graphbolt/link_prediction.py` runs a bit slower. Please keep an eye on the regression results.

**Review by Rhett-Ying — APPROVED** [review]
